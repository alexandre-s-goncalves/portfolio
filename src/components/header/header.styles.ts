import styled from 'styled-components';
import {Link} from 'react-router-dom';
import {colors} from 'resources/colors';
import {Text} from 'components/ui/text';
import {IconButton} from 'components/iconButton';
import {margin} from 'resources/margins';

export interface ExtraProps {
  $themeDark?: boolean;
  $isActivated?: boolean;
}

export const ThemeToggleButton = styled(IconButton).attrs<ExtraProps>(({$themeDark}) => ({
  iconColor: $themeDark ? colors.background01 : colors.black,
  iconWidth: 24,
  iconHeight: 24,
  cursor: 'pointer',
}))`
  background: none;
  padding: ${margin.xsmall}px;
`;

export const LogoButton = styled(IconButton).attrs<ExtraProps>(({$isActivated}) => ({
  iconWidth: 40,
  iconHeight: 40,
}))<ExtraProps>`
  background: transparent;
  cursor: ${({$isActivated}) => ($isActivated ? 'default' : 'pointer')};
  pointer-events: ${({$isActivated}) => ($isActivated ? 'none' : 'auto')};

  &:hover {
    opacity: ${({$isActivated}) => ($isActivated ? 1 : 0.8)};
  }
`;

export const TextHeader = styled(Text).attrs<ExtraProps>(({$themeDark, $isActivated}) => ({
  size: 14,
  type: 'h3',
  color: $isActivated
    ? $themeDark
      ? colors.neutralXLight
      : colors.black
    : $themeDark
      ? colors.grayLight
      : colors.darkerCadetGrey,
  marginTop: 0,
  cursor: $isActivated ? 'default' : 'pointer',
  weight: $isActivated ? 'semi-bold' : 'medium',
}))`
  transition: color 0.3s ease;

  &:hover {
    color: ${({$themeDark}) => ($themeDark ? colors.background01 : colors.black)};
  }
`;

export const HeaderContainer = styled.header<ExtraProps>`
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: ${margin.xsmall}px 120px;
  box-sizing: border-box;
  border-bottom: 1px solid ${({$themeDark}) => ($themeDark ? colors.gunmetal : colors.neutralLight)};
`;

export const Nav = styled.nav`
  display: flex;
  gap: ${margin.large}px;
`;

export const NavLink = styled(Link)`
  text-decoration: none;
`;

export const ButtonsContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  gap: ${margin.small}px;
`;
