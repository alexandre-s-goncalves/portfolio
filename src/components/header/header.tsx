import React from 'react';
import {useTheme} from 'context/ThemeContext';
import {useTranslation} from 'react-i18next';
import {useNavigate, useLocation} from 'react-router-dom';
import {namespaces} from 'utils/i18n/i18n.constants';
import {Dropdown} from 'components/dropdown';
import MoonSVG from 'assets/icons/iMoon.svg';
import SunSVG from 'assets/icons/iSun.svg';
import LogoDarkSVG from 'assets/image/logo-dark.svg';
import LogoClearSVG from 'assets/image/logo-clear.svg';
import * as S from './header.styles';

export const Header = () => {
  const {themeDark, handleToggle} = useTheme();
  const {t} = useTranslation(namespaces.pages.header);
  const navigate = useNavigate();
  const location = useLocation();

  const navigation = [
    {name: t('home'), href: '/'},
    {name: t('skills'), href: '/skills'},
    {name: t('projects'), href: '/projects'},
    {name: t('about'), href: '/about'},
  ];

  const isActive = (path: string): boolean => {
    return location.pathname === path;
  };

  const isHome = location.pathname === '/';

  const handleLogoClick = () => {
    if (!isHome) {
      navigate('/');
    }
  };

  return (
    <S.HeaderContainer $themeDark={themeDark}>
      <S.LogoButton
        testId={themeDark ? 'logo-clear' : 'logo-dark'}
        icon={themeDark ? LogoClearSVG : LogoDarkSVG}
        $isActivated={isHome}
        onClick={handleLogoClick}
        aria-label="Navigate to home"
      />
      <S.Nav>
        {navigation.map(item => (
          <S.NavLink
            key={item.href}
            to={item.href}
            data-testid={`navlink-${item.name.toLowerCase()}`}>
            <S.TextHeader $themeDark={themeDark} $isActivated={isActive(item.href)}>
              {item.name}
            </S.TextHeader>
          </S.NavLink>
        ))}
      </S.Nav>
      <S.ButtonsContainer>
        <Dropdown />
        <S.ThemeToggleButton
          testId={themeDark ? 'theme-toggle-sun' : 'theme-toggle-moon'}
          icon={themeDark ? SunSVG : MoonSVG}
          $themeDark={themeDark}
          onClick={handleToggle}
        />
      </S.ButtonsContainer>
    </S.HeaderContainer>
  );
};
