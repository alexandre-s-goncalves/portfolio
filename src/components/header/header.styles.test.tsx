import React from 'react';
import {render} from '@testing-library/react';
import {BrowserRouter} from 'react-router-dom';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import * as S from './header.styles';

describe('header.styles', () => {
  describe('HeaderContainer', () => {
    test('should match the snapshot', () => {
      const {container} = render(<S.HeaderContainer $themeDark={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should render as header element', () => {
      const {container} = render(<S.HeaderContainer $themeDark={false} />);
      const element = container.firstChild as HTMLElement;
      expect(element.nodeName).toBe('HEADER');
    });

    test('should apply light theme border color', () => {
      const {getByTestId} = render(<S.HeaderContainer data-testid="header" $themeDark={false} />);
      const header = getByTestId('header');
      expect(header).toHaveStyle('border-bottom: 1px solid #D9DDDB');
    });

    test('should apply dark theme border color', () => {
      const {getByTestId} = render(<S.HeaderContainer data-testid="header" $themeDark={true} />);
      const header = getByTestId('header');
      expect(header).toHaveStyle('border-bottom: 1px solid #323741');
    });

    test('should have full width', () => {
      const {getByTestId} = render(<S.HeaderContainer data-testid="header" $themeDark={false} />);
      const header = getByTestId('header');
      expect(header).toHaveStyle('width: 100%');
    });

    test('should have flex layout', () => {
      const {getByTestId} = render(<S.HeaderContainer data-testid="header" $themeDark={false} />);
      const header = getByTestId('header');
      expect(header).toHaveStyle('display: flex');
      expect(header).toHaveStyle('justify-content: space-between');
      expect(header).toHaveStyle('align-items: center');
    });

    test('should have correct padding', () => {
      const {getByTestId} = render(<S.HeaderContainer data-testid="header" $themeDark={false} />);
      const header = getByTestId('header');
      expect(header).toHaveStyle('padding: 8px 96px');
    });
  });

  describe('Nav', () => {
    test('should match the snapshot', () => {
      const {container} = render(<S.Nav />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should render as nav element', () => {
      const {container} = render(<S.Nav />);
      const element = container.firstChild as HTMLElement;
      expect(element.nodeName).toBe('NAV');
    });

    test('should have flex layout with gap', () => {
      const {getByTestId} = render(<S.Nav data-testid="nav" />);
      const nav = getByTestId('nav');
      expect(nav).toHaveStyle('display: flex');
      expect(nav).toHaveStyle('gap: 32px');
    });
  });

  describe('NavLink', () => {
    test('should match the snapshot', () => {
      const {container} = render(
        <BrowserRouter>
          <S.NavLink to="/">Home</S.NavLink>
        </BrowserRouter>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should render as anchor element', () => {
      const {getByText} = render(
        <BrowserRouter>
          <S.NavLink to="/">Home</S.NavLink>
        </BrowserRouter>,
      );
      const link = getByText('Home');
      expect(link.nodeName).toBe('A');
    });

    test('should have no text decoration', () => {
      const {getByText} = render(
        <BrowserRouter>
          <S.NavLink to="/">Home</S.NavLink>
        </BrowserRouter>,
      );
      const link = getByText('Home');
      expect(link).toHaveStyle('text-decoration: none');
    });

    test('should have correct href attribute', () => {
      const {getByText} = render(
        <BrowserRouter>
          <S.NavLink to="/skills">Skills</S.NavLink>
        </BrowserRouter>,
      );
      const link = getByText('Skills');
      expect(link).toHaveAttribute('href', '/skills');
    });
  });

  describe('ButtonsContainer', () => {
    test('should match the snapshot', () => {
      const {container} = render(<S.ButtonsContainer />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should render as div element', () => {
      const {container} = render(<S.ButtonsContainer />);
      const element = container.firstChild as HTMLElement;
      expect(element.nodeName).toBe('DIV');
    });

    test('should have flex layout', () => {
      const {getByTestId} = render(<S.ButtonsContainer data-testid="buttons" />);
      const container = getByTestId('buttons');
      expect(container).toHaveStyle('display: flex');
      expect(container).toHaveStyle('justify-content: center');
      expect(container).toHaveStyle('align-items: center');
    });

    test('should have correct gap', () => {
      const {getByTestId} = render(<S.ButtonsContainer data-testid="buttons" />);
      const container = getByTestId('buttons');
      expect(container).toHaveStyle('gap: 16px');
    });
  });

  describe('ThemeToggleButton', () => {
    test('should match the snapshot', () => {
      const MockIcon = () => <svg />;
      const {container} = render(<S.ThemeToggleButton icon={MockIcon} $themeDark={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should apply light theme icon color', () => {
      const MockIcon = (props: any) => <svg data-testid="icon" {...props} />;
      const {getByTestId} = render(<S.ThemeToggleButton icon={MockIcon} $themeDark={false} />);
      const icon = getByTestId('icon');
      expect(icon).toHaveAttribute('fill', '#000000');
    });

    test('should apply dark theme icon color', () => {
      const MockIcon = (props: any) => <svg data-testid="icon" {...props} />;
      const {getByTestId} = render(<S.ThemeToggleButton icon={MockIcon} $themeDark={true} />);
      const icon = getByTestId('icon');
      expect(icon).toHaveAttribute('fill', '#ffffff');
    });

    test('should have correct icon dimensions', () => {
      const MockIcon = (props: any) => <svg data-testid="icon" {...props} />;
      const {getByTestId} = render(<S.ThemeToggleButton icon={MockIcon} $themeDark={false} />);
      const icon = getByTestId('icon');
      expect(icon).toHaveAttribute('width', '24');
      expect(icon).toHaveAttribute('height', '24');
    });
  });

  describe('LogoButton', () => {
    test('should match the snapshot', () => {
      const MockIcon = () => <svg />;
      const {container} = render(<S.LogoButton icon={MockIcon} $isActivated={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should have pointer cursor when not activated', () => {
      const MockIcon = () => <svg />;
      const {getByTestId} = render(
        <S.LogoButton data-testid="logo" icon={MockIcon} $isActivated={false} />,
      );
      const button = getByTestId('logo');
      expect(button).toHaveStyle('cursor: pointer');
    });

    test('should have default cursor when activated', () => {
      const MockIcon = () => <svg />;
      const {getByTestId} = render(
        <S.LogoButton data-testid="logo" icon={MockIcon} $isActivated={true} />,
      );
      const button = getByTestId('logo');
      expect(button).toHaveStyle('cursor: default');
    });

    test('should have pointer-events none when activated', () => {
      const MockIcon = () => <svg />;
      const {getByTestId} = render(
        <S.LogoButton data-testid="logo" icon={MockIcon} $isActivated={true} />,
      );
      const button = getByTestId('logo');
      expect(button).toHaveStyle('pointer-events: none');
    });

    test('should have pointer-events auto when not activated', () => {
      const MockIcon = () => <svg />;
      const {getByTestId} = render(
        <S.LogoButton data-testid="logo" icon={MockIcon} $isActivated={false} />,
      );
      const button = getByTestId('logo');
      expect(button).toHaveStyle('pointer-events: auto');
    });

    test('should have correct icon dimensions', () => {
      const MockIcon = (props: any) => <svg data-testid="icon" {...props} />;
      const {getByTestId} = render(<S.LogoButton icon={MockIcon} $isActivated={false} />);
      const icon = getByTestId('icon');
      expect(icon).toHaveAttribute('width', '48');
      expect(icon).toHaveAttribute('height', '48');
    });
  });

  describe('TextHeader', () => {
    test('should match the snapshot with active state', () => {
      const {container} = render(<S.TextHeader $themeDark={false} $isActivated={true} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match the snapshot with inactive state', () => {
      const {container} = render(<S.TextHeader $themeDark={false} $isActivated={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should apply active color in light theme', () => {
      const {getByTestId} = render(
        <S.TextHeader data-testid="text" $themeDark={false} $isActivated={true} />,
      );
      const text = getByTestId('text');
      expect(text).toHaveStyle('color: #000000');
    });

    test('should apply inactive color in light theme', () => {
      const {getByTestId} = render(
        <S.TextHeader data-testid="text" $themeDark={false} $isActivated={false} />,
      );
      const text = getByTestId('text');
      expect(text).toHaveStyle('color: #8f95a3');
    });

    test('should apply active color in dark theme', () => {
      const {getByTestId} = render(
        <S.TextHeader data-testid="text" $themeDark={true} $isActivated={true} />,
      );
      const text = getByTestId('text');
      expect(text).toHaveStyle('color: #ffffff');
    });

    test('should apply inactive color in dark theme', () => {
      const {getByTestId} = render(
        <S.TextHeader data-testid="text" $themeDark={true} $isActivated={false} />,
      );
      const text = getByTestId('text');
      expect(text).toHaveStyle('color: #b9bcc4');
    });

    test('should have pointer cursor when not activated', () => {
      const {getByTestId} = render(
        <S.TextHeader data-testid="text" $themeDark={false} $isActivated={false} />,
      );
      const text = getByTestId('text');
      expect(text).toHaveStyle('cursor: pointer');
    });

    test('should have default cursor when activated', () => {
      const {getByTestId} = render(
        <S.TextHeader data-testid="text" $themeDark={false} $isActivated={true} />,
      );
      const text = getByTestId('text');
      expect(text).toHaveStyle('cursor: default');
    });
  });
});
