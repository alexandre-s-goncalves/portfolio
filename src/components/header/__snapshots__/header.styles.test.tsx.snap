// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header.styles ButtonsContainer should match the snapshot 1`] = `
.c0 {
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 16px;
}

<div
  class="c0"
/>
`;

exports[`header.styles HeaderContainer should match the snapshot 1`] = `
.c0 {
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 8px 120px;
  box-sizing: border-box;
  border-bottom: 1px solid #D9DDDB;
}

<header
  class="c0"
/>
`;

exports[`header.styles LogoButton should match the snapshot 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  outline: none;
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.c0:hover:not(:disabled) {
  opacity: 0.8;
}

.c0:active:not(:disabled) {
  transform: scale(0.95);
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c0:focus-visible {
  outline: 2px solid rgba(0, 0, 0, 0.3);
  outline-offset: 2px;
}

.c1 {
  background: transparent;
  cursor: pointer;
  pointer-events: auto;
}

.c1:hover {
  opacity: 0.8;
}

<button
  class="c0 c1"
  data-testid="icon-button"
>
  <svg />
</button>
`;

exports[`header.styles Nav should match the snapshot 1`] = `
.c0 {
  display: flex;
  gap: 32px;
}

<nav
  class="c0"
/>
`;

exports[`header.styles NavLink should match the snapshot 1`] = `
.c0 {
  text-decoration: none;
}

<a
  class="c0"
  data-discover="true"
  href="/"
>
  Home
</a>
`;

exports[`header.styles TextHeader should match the snapshot with active state 1`] = `
.c0 {
  font-size: 14px;
  font-family: Inter-Regular;
  color: #000000;
  margin-top: 0px;
  font-weight: normal;
  text-transform: none;
  cursor: default;
}

.c1 {
  transition: color 0.3s ease;
}

.c1:hover {
  color: #000000;
}

<h3
  class="c0 c1"
/>
`;

exports[`header.styles TextHeader should match the snapshot with inactive state 1`] = `
.c0 {
  font-size: 14px;
  font-family: Inter-Regular;
  color: #8f95a3;
  margin-top: 0px;
  font-weight: normal;
  text-transform: none;
  cursor: pointer;
}

.c1 {
  transition: color 0.3s ease;
}

.c1:hover {
  color: #000000;
}

<h3
  class="c0 c1"
/>
`;

exports[`header.styles ThemeToggleButton should match the snapshot 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  outline: none;
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.c0:hover:not(:disabled) {
  opacity: 0.8;
}

.c0:active:not(:disabled) {
  transform: scale(0.95);
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c0:focus-visible {
  outline: 2px solid rgba(0, 0, 0, 0.3);
  outline-offset: 2px;
}

.c1 {
  background: none;
  padding: 8px;
}

<button
  class="c0 c1"
  cursor="pointer"
  data-testid="icon-button"
>
  <svg />
</button>
`;
